akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala

```diff
@@ -450,9 +450,12 @@ private[cluster] final class ClusterDomainEventPublisher extends Actor with Acto
   }
 
   def publishDiff(oldState: MembershipState, newState: MembershipState, pub: AnyRef ⇒ Unit): Unit = {
+    def inSameDc(reachabilityEvent: ReachabilityEvent): Boolean =
+      reachabilityEvent.member.dataCenter == selfDc
+
     diffMemberEvents(oldState, newState) foreach pub
-    diffUnreachable(oldState, newState) foreach pub
-    diffReachable(oldState, newState) foreach pub
+    diffUnreachable(oldState, newState) filter inSameDc foreach pub
+    diffReachable(oldState, newState) filter inSameDc foreach pub
```

Comment:
um, that is not enough :)

In the end you should find `Reachability.Record` entries with `observer.dataCenter == selfDc && subject.dataCenter == selfDc`. All other records should be thrown away.

Perhaps something like this in `MembershipState`:
```
dcReachability.remove(members.collect { case m if m.dataCenter != selfDc ⇒ m.uniqueAddress })
```